### ISSUE TITLE
LDAP authentication should return a meaningful error when the LDAP server is unavailable 

### ISSUE THREAD

**Issue by Jayfrown** [symfony/symfony#44089]
### Symfony version(s) affected

5.3

### Description

In case the configured LDAP server is not currently reachable, the `json_login_ldap` authenticator returns this to the user:

```
{
    "code": 401,
    "message": "Invalid credentials."
}
```

Internally, a PHP notice is logged:

```
Warning: ldap_bind(): Unable to bind to server: Can't contact LDAP server
```

and `symfony/ldap` throws `Symfony\Component\Security\Core\Exception\BadCredentialsException`, which doesn't seem quite right.

I think it would make more sense for the server to return a `500` stating something went wrong while attempting to verify credentials.

### How to reproduce

- Have a working LDAP authentication setup
- Introduce connectivity issues (firewall, stop the ldap server, what have you)
- Try to authenticate using valid credentials

### Possible Solution

When `ldap_bind()` finds it is `Unable to bind to server: Can't contact LDAP server`, instead of throwing a `BadCredentialsException` we could throw something else to indicate there was an issue while attempting to verify credentials. Various authenticators (like `json_login_ldap` could then return meaningful errors to the user.

**Comment by derrabus** [comment]
Would you like to work on a fix?

**Comment by Jayfrown** [comment]
Yeah, I would love to contribute. I do think I would need some pointers on where to start.

**Comment by derrabus** [comment]
Pointers:

* Check out the 5.3 branch of `symfony/symfony`.
* Look into the LDAP component and find the `ldap_bind()` call.
* Find the corresponding test for that piece of code.
* Add a test that reproduces your problem.
* Fix the code so your new test passes.

**Comment by Jayfrown** [comment]
Thanks @derrabus. Current tests seem to rely on a mocked Adapter / Connection, and I cannot quite figure out how to add a failing test for my case.

I did some further digging however.

`Adapter/ExtLdap/Connection.php` starting from line 76:

```php
          if (false === @ldap_bind($this->connection, $dn, $password)) {
              $error = ldap_error($this->connection);
              switch (ldap_errno($this->connection)) {
                  case self::LDAP_INVALID_CREDENTIALS:
                      throw new InvalidCredentialsException($error);
                  case self::LDAP_TIMEOUT:
                      throw new ConnectionTimeoutException($error);
                  case self::LDAP_ALREADY_EXISTS:
                      throw new AlreadyExistsException($error);
              }
              throw new ConnectionException($error);
          }
```
 
In the case of an unreachable LDAP server, printing the results of `ldap_errno()`/`ldap_error()` yields `-1`/`Can't contact LDAP server`, so the LDAP component appropriately throws a `ConnectionException()` in that case.

However, `ldap/Security/CheckLdapCredentialsListener.php` catches the `ConnectionException` and throws a `BadCredentialsException` at lines 102/103. This was introduced in commit `20962e604a7`, meaning it has been there ever since the Authenticator system started supporting LDAP.

Removing the try/catch yields the 500 ISE that I had initially expected. However, it might be better for the authenticator to still return a HTTP 401 with a JSON response, just with a different message. I would propose the following:

```json
{
    "code": 401,
    "message": "Can't contact LDAP server"
}
````

Thoughts?

---

Tagging @wouterj as they're the author of `20962e604a7` and might have an opinion on this as well.

**Comment by derrabus** [comment]
> ```json
> {
>     "code": 401,
>     "message": "Can't contact LDAP server"
> }
> ```
> 
> Thoughts?

I'm pretty certain we don't want to tell the client that the LDAP server is unreachable, espacially not in prod mode. Writing an error to the log is more appropriate, I think.

But I'd also be eager to hear @wouterj's input on this topic. 🙂

**Comment by Jayfrown** [comment]
Yes, maybe a message like `"The provided credentials could not be validated."` would make more sense. Currently it behaves as if the credentials are known to be invalid.

**Comment by wouterj** [comment]
We normally don't catch these type of errors in authenticators (e.g. you get a 500 if the db is down).

However, for LDAP a connection error of `bind()` can mean 3 things:
1. The DN (username) is incorrect
2. The password is incorrect
3. The LDAP server is unreachable

https://github.com/symfony/symfony/blob/54015cccf0236bbdb38cd5abae5b2bdc89aa8ac2/src/Symfony/Component/Ldap/LdapInterface.php#L28-L33

We check (1) using an LDAP query before and we check (2) using the try...catch statement. This is why the message is like it is.

I've never been able to set-up an ldap server, so if anyone has LDAP knowledge and knows how to better solve this, please be welcome to submit a proposal :)

**Comment by Jayfrown** [comment]
Thanks for your input @wouterj 

> We normally don't catch these type of errors in authenticators (e.g. you get a 500 if the db is down).

Ah yes, you're right. In that case it would make sense to return a 500 if LDAP is down as well.

> [..] and we check (2) using the try...catch statement

Well, the try...catch statement explicitly catches a `ConnectionException` and throws a `BadCredentialsException`, however if I understand correctly the LDAP component would throw an `InvalidCredentialsException` as outlined above.

~~I am confused about the behavior of these exceptions being thrown. If the `LdapUserProvider` catches the `ConnectionException`, the `InvalidCredentialsException` seems to be suppressed, but if I remove the catch for `ConnectionException`, I see the `InvalidCredentialsException` being thrown, resulting in a 500 ISE.~~

~~I thought if the switch case results in throwing `InvalidCredentialsException`, execution would stop from there, but this does not appear to be the case. Maybe someone else knows what's going on there?~~

I was being dumb, `class InvalidCredentialsException extends ConnectionException`.

---

After removing the try...catch for `ConnectionException` in `ldap/Security/CheckLdapCredentialsListener.php` and `ldap/Security/LdapUserProvider.php` I have tested the following scenarios:

<details>
<summary>LDAP server unreachable</summary>

`Adapter/ExtLdap/Connection.php` throws a `ConnectionException` which is not caught, and this results in a 500 ISE shown to the user. In development mode, you see a relevant trace and the correct error hinting to an unreachable LDAP server, and in production mode this wouldn't be shown to the user, and would be logged.

I believe this both replicates behavior of other authenticators (as you said, you get a 500 if the DB is down) and should also satisfy @derrabus comment about not telling the client that LDAP is unreachable in production mode. This would only happen in development mode, where you see the stack trace resulting in the 500 ISE.
</details>

<details>
<summary>LDAP server reachable, with invalid search bind credentials</summary>

The `LdapUserProvider` tries to search the LDAP directory for the provided user. To search, it tries to authenticate using the (invalid) bind credentials, resulting in `Adapter/ExtLdap/Connection.php` throwing an `InvalidCredentialsException` which is not caught. As this is indicative of a misconfiguration within the application (ie. the app's search bind credentials are wrong, and searching for any user will always fail) I believe a 500 ISE is appropriate.

We could opt to catch the `InvalidCredentialsException` inside the `LdapUserProvider` and throw something more specific, to indicate it is the applications bind credentials that are invalid, as opposed to the credentials provided during authentication.
</details>

<details>
<summary>LDAP server reachable, with valid search bind credentials, authenticating as an unknown user</summary>

As the search bind credentials are valid, the `LdapUserProvider` searches the LDAP directory, finds no entries, and throws a `UserNotFoundException` which results in the authenticator returning `{"code":401,"message":"Invalid credentials."}`. I believe this is good / expected behavior.
</details>

<details>
<summary>LDAP server reachable, with valid search bind credentials, using known user with invalid password</summary>

The `ldap/Adapter/ExtLdap/Connection.php` throws an `InvalidCredentialsException` which is not caught, resulting in a 500 ISE. I believe the `ldap/Security/CheckLdapCredentialsListener.php` should catch the `InvalidCredentialsException` (instead of the `ConnectionException` it currently catches) and throw a `BadCredentialsException` such that the authenticator returns `{"code":401,"message":"Invalid credentials."}` in this case.
</details>

<details>
<summary>LDAP server reachable, with valid search bind credentials, using known user with valid password</summary>

The credentials are validated and the user is logged in, everything works as expected
</details>

---

In conclusion I believe we should refrain from catching the `ConnectionException` within the `LdapUserProvider` and `CheckLdapCredentialsListener`. The `LdapUserProvider` still appears to function as it should, and the `CheckLdapCredentialsListener` should catch the `InvalidCredentialsException` instead.

This would result in the application returning a 500 ISE in the following scenarios:

- If the LDAP server in unreachable
- If the application's LDAP search bind credentials are invalid

And returning a 401 "invalid credentials" if the provided credentials are known to be invalid.

I have opened https://github.com/symfony/symfony/pull/44896 with proposed changes. I'm not sure of the target branch.
